sycl/test-e2e/Annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,121 @@
+  assert(sycl::get_pointer_type(APtr2.get(), Ctx) == alloc::device);
+
+  auto APtr3 = malloc_host_annotated<int>(N, q);
+  assert(sycl::get_pointer_type(APtr3.get(), Ctx) == alloc::host);
+
+  if (dev.has(sycl::aspect::usm_shared_allocations)) {
+    auto APtr4 = malloc_shared_annotated(N, q);
+    assert(sycl::get_pointer_type(APtr4.get(), Ctx) == alloc::shared);
+    free(APtr4, q);
+  }
+
+  // the USM kind is specified in the propList
+  properties P2{conduit, cache_config{large_slm}};
+  auto APtr5 = malloc_annotated<int>(N, q, alloc::device, P2);
+  assert(sycl::get_pointer_type(APtr5.get(), Ctx) == alloc::device);
+
+  // the USM kind is specified in the propList
+  properties P3{usm_kind<alloc::host>};
+  auto APtr6 = malloc_annotated<int>(N, q, P3);
+  assert(sycl::get_pointer_type(APtr6.get(), Ctx) == alloc::host);
+
+  free(APtr1, q);
+  free(APtr2, q);
+  free(APtr3, q);
+  free(APtr5, q);
+  free(APtr6, q);
+}
+
+void TestAlign(sycl::queue &q) {
+  properties P7{alignment<512>};
+  // The raw pointer of APtr1 is 512-byte aligned
+  auto APtr1 = malloc_device_annotated<int>(N, q, P7);
+  assert(((uintptr_t)APtr1.get() & 511) == 0);
+
+  properties P8{alignment<2>};
+  // The raw pointer of APtr2 is sizeof(double)-byte aligned, e.g., 8 for some
+  // implementations
+  auto APtr2 = malloc_device_annotated<double>(N, q, P8);
+  assert(((uintptr_t)APtr2.get() & 7) == 0);
+
+  properties P9{alignment<64>};
+  // The raw pointer of APtr3 is 64-byte aligned
+  auto APtr3 = malloc_device_annotated(N, q, P9);
+  assert(((uintptr_t)APtr3.get() & 63) == 0);
+
+  // The raw pointer of APtr4 is 1024-byte aligned
+  // Note: APtr4 does not have the alignment
+  // property. The alignment is runtime information.
+  auto APtr4 = aligned_alloc_device_annotated<int>(
+      1024 /* alignment */, N, q,
+      sycl::ext::oneapi::experimental::detail::empty_properties_t{});
+  assert(((uintptr_t)APtr4.get() & 1023) == 0);
+
+  auto APtr41 = aligned_alloc_device_annotated(1024 /* alignment */, N, q);
+  assert(((uintptr_t)APtr41.get() & 1023) == 0);
+
+  // The raw pointer of APtr5 is 64-byte aligned
+  auto APtr5 = aligned_alloc_device_annotated<int>(64, N, q, P9);
+  assert(((uintptr_t)APtr5.get() & 63) == 0);
+
+  // The raw pointer of APtr6 is 128-byte
+  // aligned Note: APtr15 has the alignment property with value 64, because this
+  // is the alignment known at compile-time
+  auto APtr6 = aligned_alloc_device_annotated<int>(128, N, q, P9);
+  assert(((uintptr_t)APtr6.get() & 127) == 0);
+
+  // The raw pointer of APtr7 is 64-byte aligned
+  auto APtr7 = aligned_alloc_device_annotated<int>(16, N, q, P9);
+  assert(((uintptr_t)APtr7.get() & 63) == 0);
+
+  properties P10{alignment<8>};
+  // The raw pointer of APtr8 is 56-byte
+  // aligned (if this alignment is supported by the implementation)
+  auto APtr8 = aligned_alloc_device_annotated<int>(7, N, q, P10);
+  assert(((uintptr_t)APtr8.get() % 56) == 0);
+
```

Comment:
Can you add the following alignment test cases:

1. alignment < k > where k is larger than sizeof(T), but not a multiple of sizeof(T)
2. alignment <0>
3. alignment<0> with a runtime alignment of k != 0
4. alignment < k > with a runtime alignment of 0

Reply:
In the first case, the resulting alignment should be the LCM of sizeof(T) and k.
As mentioned in the PR description, the resulting alignment should be k in both cases 3 and 4.
What happens if alignment<0> is specified but no other non-zero alignment is specified? I don't see a value of 0 discussed in the annotated_ptr spec: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc
@tiwaria1 do you know how an alignment of 0 should be handled?

Reply:
I think it should be a compile time error to specify an alignment of 0. Would you agree?

Reply:
I looked at the implementation of the existing USM malloc APIs, e.g. `malloc_device` calls `aligned_alloc_device(0, ...)` under the hood, so 0 seems like a valid argument

Reply:
That doesn't make sense - maybe it is ignored somewhere in the stack then because alignment of 0 is meaningless.

Reply:
I think it makes sense to allow `alignment<0>` without reporting an error, and have it mean that no alignment is specified.
Given that it's up to each implementation to define which alignment values are valid, I think this is our decision to make in the implementation.
However, we should document this to our users.